File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/facade/FacadeSystem.java
package jv.com.tencent.cain.design_pattern.facade;

/**
 * @author : jiangyu
 * @date : 2020/2/5
 * @desc : xxx
 */
public class FacadeSystem {
    private SubSystemOne subSystemOne;
    private SubSystemTwo subSystemTwo;
    private SubSystemThree subSystemThree;

    public FacadeSystem() {
        subSystemOne = new SubSystemOne();
        subSystemTwo = new SubSystemTwo();
        subSystemThree = new SubSystemThree();
    }

    public void methodA() {
        System.out.println("---方法组A---");
        subSystemOne.methodOne();
        subSystemTwo.methodTwo();
        System.out.println();
    }

    public void methodB() {
        System.out.println("---方法组B---");
        subSystemOne.methodOne();
        subSystemThree.methodThree();
        System.out.println();
    }

    public void methodC() {
        System.out.println("---方法组C---");
        subSystemTwo.methodTwo();
        subSystemThree.methodThree();
        System.out.println();
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/facade/FacadeSystem.kt
package kt.com.tencent.cain.desigin_pattern.facade

/**
 * @author : jiangyu
 * @date   : 2020/2/5
 * @desc   : xxx
 */
class FacadeSystem {
    val subSystemOne: SubSystemOne by lazy { SubSystemOne() }
    val subSystemTwo: SubSystemTwo by lazy { SubSystemTwo() }
    val subSystemThree: SubSystemThree by lazy { SubSystemThree() }

    fun funcA() {
        println("---方法组A---")
        subSystemOne.funcOne()
        subSystemTwo.funcTwo()
        println()
    }

    fun funcB() {
        println("---方法组B---")
        subSystemOne.funcOne()
        subSystemThree.funcThree()
        println()
    }

    fun funcC() {
        println("---方法组C---")
        subSystemTwo.funcTwo()
        subSystemThree.funcThree()
        println()
    }
}